Add `KVM_CAP_GUEST_DEBUG_HW_BPS/WPS`

michael2012z opened the pull request:
### Summary of the PR

The API `KVM_CAP_GUEST_DEBUG_HW_BPS` and `KVM_CAP_GUEST_DEBUG_HW_WPS`
are required to determine the number of supported guest debug registers.

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] Any newly added `unsafe` code is properly documented.

michael2012z committed:
Add `KVM_CAP_GUEST_DEBUG_HW_BPS/WPS`

The API `KVM_CAP_GUEST_DEBUG_HW_BPS` and `KVM_CAP_GUEST_DEBUG_HW_WPS`
are required to determine the number of supported guest debug registers.

Signed-off-by: Michael Zhao <michael.zhao@arm.com>

michael2012z commented on src/ioctls/system.rs:
```diff
@@ -584,6 +600,14 @@ mod tests {
+    #[cfg(target_arch = "aarch64")]
+    fn test_guest_debug_hw_capacity() {
+        let kvm = Kvm::new().unwrap();
+        assert!(kvm.get_guest_debug_hw_bps() >= 0);
+        assert!(kvm.get_guest_debug_hw_wps() >= 0);
```
To discuss:

I am not sure about the rule to check the result. The returned value vary on different machines. ">=0" seems not very meaningful.

And good idea?

andreeaflorescu replied:
I think it's okay to check it as it is now in the test because we don't know what hosts we're running on. Can you add a comment about why we're doing the check this way? (i.e. different values on different platforms).

michael2012z commented on src/ioctls/system.rs:
```diff
@@ -97,11 +97,27 @@ impl Kvm {
+    ///
+    /// Returns 0 if the capability is not available and a positive integer otherwise.
+    #[cfg(target_arch = "aarch64")]
+    pub fn get_guest_debug_hw_wps(&self) -> i32 {
```
If there is increasing requirements to call the APIs that return a integer, is it good to expose `check_extension_int` itself?

andreeaflorescu replied:
We can mark it as public as well, I don't have any concern with that. That's a good point!

michael2012z replied:
I added a commit for that.

andreeaflorescu commented on src/ioctls/system.rs:
```diff
@@ -97,11 +97,27 @@ impl Kvm {
 
+    /// AArch64 specific call to get the number of supported hardware breakpoints.
+    ///
+    /// Returns 0 if the capability is not available and a positive integer otherwise.
```
Does the integer bear any meaning? Can we return a bool in case it doesn't have any useful information?

rbradford replied:
The integer is the number of breakpoints.

michael2012z replied:
The integer is useful in supporting the debug of guest kernel. 
On x86, the max number of breakpoints is 4, while on arm the number is not fixed but is indicated by this API.

michael2012z committed:
Expose `Kvm::check_extension_int()`

The function is useful for checking the capabilities that return numbers
rather than booleans.

Signed-off-by: Michael Zhao <michael.zhao@arm.com>

acatangiu reviewed (approved):
Looks good to me!
